hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java

```diff
@@ -1809,12 +1819,30 @@ else if ( property.isAnnotationPresent( OneToOne.class ) ) {
 				}
 
 				//FIXME support a proper PKJCs
-				boolean trueOneToOne = property.isAnnotationPresent( PrimaryKeyJoinColumn.class )
+				final boolean hasPkjc = property.isAnnotationPresent( PrimaryKeyJoinColumn.class )
 						|| property.isAnnotationPresent( PrimaryKeyJoinColumns.class );
+				boolean trueOneToOne = hasPkjc;
 				Cascade hibernateCascade = property.getAnnotation( Cascade.class );
 				NotFound notFound = property.getAnnotation( NotFound.class );
 				boolean ignoreNotFound = notFound != null && notFound.action().equals( NotFoundAction.IGNORE );
+				// MapsId means the columns belong to the pk;
+				// A @MapsId association (obviously) must be non-null when the entity is first persisted.
+				// If a @MapsId association is not mapped with @NotFound(IGNORE), then the association
+				// is mandatory (even if the association has optional=true).
+				// If a @MapsId association has optional=true and is mapped with @NotFound(IGNORE) then
+				// the association is optional.
+				// @OneToOne(optional = true) with @PKJC makes the association optional.
+				final boolean mandatory =
+						!ann.optional() ||
+								property.isAnnotationPresent( Id.class ) ||
+								( property.isAnnotationPresent( MapsId.class ) && !ignoreNotFound );
+				// An optional @OneToOne with @PKJC or @PKJCs has an implicit @NotFound(IGNORE)
+				// (see HHH-4982).
+				if ( hasPkjc && !mandatory ) {
+					ignoreNotFound = true;
```

Comment:
Perhaps this workaround is not needed after the alternative fix for the `isNullable()` method?